### PR TITLE
Feature: build cluster image with buildah

### PR DESCRIPTION
**Pull request by justadogistaken**
### Describe what this PR does / why we need it
Sealer has integrated the buildah to build a cluster image.
But There are some todos.
1. How to save container images?
solution: scan the whole kubefile, pre-download all "http://" to build context. And check the `image:` of them, then save then in registry under build context. Finally add an instruction `COPY registry /` before real building procedure.

2. Semantic translator.
All basic instructions are `FROM`, `ADD`, `COPY`, `RUN`. We will implement all [new instructions](https://github.com/sealerio/sealer/issues/1579) based on these basics.
So we are going to need a Semantic translator.

A brief flow as following:
sealer-build drawio

todos:
- [x] deal with the flags from buildah
- [x] commit images by buildah.
- [x] find a place for flags
- [x] find a place for global info. Like authfile
- [x] support for sealer rootfs
- [x] sealer v1.Image refactor. And annotate it to oci image.
- [x] list sealer images only for `images`. Add label to image.
- [x] clean environment after build.
- [ ] multiple platform support.

### CODE REVIEW

**Comment by justadogistaken** [comment]
Maybe the sealer imageSpec and image is not in need any more.
```
// ImageSpec defines the desired state of Image
type ImageSpec struct {
	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
	// Important: Run "make" to regenerate code after modifying this file

	// Foo is an example field of Image. Edit Image_types.go to remove/update
	ID            string      `json:"id,omitempty"`
	Layers        []Layer     `json:"layers,omitempty"`
	SealerVersion string      `json:"sealer_version,omitempty"`
	Platform      Platform    `json:"platform"`
	ImageConfig   ImageConfig `json:"image_config"`
}
```

The ImageConfig is valuable probably.
```
type ImageConfig struct {
	// define this image is application image or normal image.
	ImageType string            `json:"image_type,omitempty"`
	Cmd       ImageCmd          `json:"cmd,omitempty"`
	Args      ImageArg          `json:"args,omitempty"`
	Labels    map[string]string `json:"labels,omitempty"`
}
```

File: cmd/sealer/cmd/build.go
Comment (allencloud) [inline cmd/sealer/cmd/build.go:62]: please do not fatal here, just return an error and making cobra framework to deal with this error is OK.

File: cmd/sealer/cmd/build.go
Comment (allencloud) [inline cmd/sealer/cmd/build.go:67]: Error without wrap is unnecessary to log here, cobra framework will handle and print it.

File: build/buildinstruction/utils.go
Comment (justadogistaken) [inline build/buildinstruction/utils.go:54]: Ignore modification of this file.

**Comment by justadogistaken** [comment]
@sealerio/sealer-maintainers @sealerio/sealer-reviewers Help me review this, plz.

File: test/testhelper/settings/common.go
Comment (justadogistaken) [inline test/testhelper/settings/common.go:81]: change it back before merging.

**Comment by justadogistaken** [comment]
related issues will be fixed.
https://github.com/sealerio/sealer/issues/1581
https://github.com/sealerio/sealer/issues/1538

**Comment by allencloud** [comment]
@justadogistaken Conflict happens. Please update this ASAP. And we will merge this with high priority.

File: apply/driver/local.go
Comment (allencloud) [inline apply/driver/local.go:168]: typo, s/contaienrs/containers

File: cmd/sealer/cmd/build.go
Comment (allencloud) [inline cmd/sealer/cmd/build.go:72]: Here you directly ignore the args, because you have not passed args into function buildSealerImage(). Then I am wondering if we could update `Args: cobra.MaximumNArgs(1),` to be `Args: cobra.NoArgs,`.

File: cmd/sealer/cmd/tag.go
Comment (allencloud) [inline cmd/sealer/cmd/tag.go:25]: create one or more tags? Since you allow to input multiple tags.

File: pkg/guest/guest.go
Comment (allencloud) [inline pkg/guest/guest.go:98]:
```
                 if len(clusterCmd) != 0 {
			return clusterCmd
		}
                 return clusterCmd
}
```
I think the condition is useless. You can directly `return clusterCmd`.

File: pkg/imageengine/buildah/build.go
Comment (allencloud) [inline pkg/imageengine/buildah/build.go:308]: I found that this function always returns a nil-error. Is it reasonable? @justadogistaken

File: pkg/imageengine/buildah/commit.go
Comment (allencloud) [inline pkg/imageengine/buildah/commit.go:78]: I suggest that you could print the detailed error reason, like `xxx candidate name is empty`? 
to add error message readability?

File: pkg/imageengine/buildah/common.go
Comment (allencloud) [inline pkg/imageengine/buildah/common.go:83]: What will happen if `name` is empty string? The code logic seems to implicit when `name` is empty string.

File: pkg/imageengine/buildah/common.go
Comment (allencloud) [inline pkg/imageengine/buildah/common.go:121]: How about naming this function `getImageType()`?

File: pkg/imageengine/buildah/copy.go
Comment (allencloud) [inline pkg/imageengine/buildah/copy.go:41]: s/specifuled/specified

File: pkg/imageengine/buildah/common.go
Comment (justadogistaken) [reply pkg/imageengine/buildah/common.go:83]: Will "return nil, errors.Errorf("error finding build container")"

File: pkg/imageengine/buildah/from.go
Comment (allencloud) [inline pkg/imageengine/buildah/from.go:164]: I suggest that we define this function like the following way:
```
func (builder *buildah.Builder) Build(quiet bool) error
```

Comment (justadogistaken) [reply pkg/imageengine/buildah/from.go:164]: This function if copied from buildah.

File: pkg/imageengine/buildah/from.go
Comment (allencloud) [inline pkg/imageengine/buildah/from.go:181]: try to use another name, like `srcs = args[:size-1]` ? although the original one has no incorrectness, different naming may bring more convenience.

File: pkg/imageengine/buildah/from.go
Comment (allencloud) [inline pkg/imageengine/buildah/from.go:238]: I think we should return an error, since unexpected command is input. And we should consider it as illegal behaviour.

File: pkg/imageengine/buildah/init.go
Comment (allencloud) [inline pkg/imageengine/buildah/init.go:62]: Quite strange for the function definition. We usually create one if it does not exist, but we seldom write it it the path does not exist. Maybe we could separate the function into two parts, createFileIfNotExist and writefile.

File: pkg/imageengine/buildah/manifest.go
Comment (allencloud) [inline pkg/imageengine/buildah/manifest.go:62]: s/on/one

File: pkg/imageengine/buildah/mount.go
Comment (allencloud) [inline pkg/imageengine/buildah/mount.go:30]: s/specifiled/specified

File: pkg/imageengine/buildah/push.go
Comment (allencloud) [inline pkg/imageengine/buildah/push.go:89]: s/on/one

**Review by allencloud — APPROVED** [review]

**Comment by allencloud** [comment]
LGTM, great work 👍🏻

**Pull request merged**
